test/serializer/additional-functions/DeadCodeCheck.js

```diff
@@ -0,0 +1,13 @@
+// does not contain:x.thisShouldNotAppear
+
+function fn(x) {
+  var a = x.thisShouldNotAppear;
```

Comment:
What if this is a getter? Why is this safe to eliminate?

Reply:
It's safe because this only happens in pure scope. Even if there was a getter, the getter should be a pure function because we're in pure scope, and not change state of that around it, so not calling it should not cause side-effects.

Reply:
Interestingly, Google Closure Compiler also applies the same optimization heuristic here too.

Reply:
All optimized functions run in pure scope. If something is doing as you mention above, then that breaks our constraint and it shouldn't be used with pure scope.

Reply:
The getter could also be:

```js
const x = {
  get thisShouldNotAppear() {
    console.log('side effect!');
    return 42;
  },
};
```

Why is this safe to eliminate?

Reply:
In pure scope, it's assumed that abstract values don't have side-effects that are unknown: https://github.com/facebook/prepack/wiki/Compiler-assumptions#pure-scope-assumptions

Reply:
> Property reads are assumed to be side effect free. If a property is backed by a getter, then the getter must be a pure function.

I see 👍

Reply:
We shouldn't follow that definition. This is not 100% correct definition and we need to fix it. We do need to support lazily initialized things. If those things mutate things we already know about and if that might have consequences.

The assumption is more limited than saying that all getters are pure. The exact definition should be more along the lines of exactly what this PR does which is "it's ok to remove redundant getters if the return value isn't used and it's ok to reuse the same value extracted from a getter multiple times and assume it is idempotent".

But even without that narrower assumption, this is *still* correct. 

Inside a pure function, you can't have side-effects. So true side-effects like logging is already not allowed. You can however mutate anything reachable from the object passed to the getter.

There are some slight nuances where we need to havoc the object as a result. However, in practice, if we are seeing an abstract getter it is because we don't know anything about the object anyway. So havocing isn't even a big deopt. However, there are theoretical cases where this can happen. E.g. the object is a subclass of an abstract class. This almost never happens in practice though.

But havocing ensures that we don't assume something about the state of the object passed in.

Since it can't have any effect on objects created outside the function, the only way any mutations in the getter can matter, is if its return value is reachable from the return value of the function.

*That* is really what `isPure` is saying.

It's not actually limited to pure getters but all these constraints add up to making it effectively the same.